porch/pkg/registry/porch/packagecommon.go

```diff
@@ -141,30 +145,71 @@ func (r *packageCommon) updatePackageRevision(ctx context.Context, name string,
 
 	oldPackage, err := r.getPackage(ctx, name)
 	if err != nil {
-		return nil, false, err
+		if apierrors.IsNotFound(err) {
+			// For server-side apply, we create new objects where they don't already exist
+		} else {
+			return nil, false, err
+		}
+	}
+
+	isCreate := false
+	var oldRuntimeObj runtime.Object // This has to be runtime.Object (and not *api.PackageRevision) or else nil-checks fail (because a nil object is not a nil interface)
+	if oldPackage != nil {
+		oldRuntimeObj = oldPackage.GetPackageRevision()
+	} else {
+		isCreate = true
 	}
 
-	oldObj := oldPackage.GetPackageRevision()
-	newRuntimeObj, err := objInfo.UpdatedObject(ctx, oldObj)
+	newRuntimeObj, err := objInfo.UpdatedObject(ctx, oldRuntimeObj)
 	if err != nil {
 		klog.Infof("update failed to construct UpdatedObject: %v", err)
 		return nil, false, err
 	}
 
-	r.updateStrategy.PrepareForUpdate(ctx, newRuntimeObj, oldObj)
+	// This type conversion is necessary because mutations work with unversioned types
+	// (mostly for historical reasons).  So the server-side-apply library returns an unversioned object.
+	if unversioned, isUnversioned := newRuntimeObj.(*unversionedapi.PackageRevision); isUnversioned {
+		klog.Warningf("converting from unversioned to versioned object")
+		typed := &api.PackageRevision{}
+		if err := r.scheme.Convert(unversioned, typed, nil); err != nil {
+			return nil, false, fmt.Errorf("failed to convert %T to %T: %w", unversioned, typed, err)
+		}
+		newRuntimeObj = typed
+	}
 
-	if updateValidation != nil {
-		err := updateValidation(ctx, newRuntimeObj, oldObj)
-		if err != nil {
-			klog.Infof("update failed validation: %v", err)
-			return nil, false, err
+	r.updateStrategy.PrepareForUpdate(ctx, newRuntimeObj, oldRuntimeObj)
+
+	if !isCreate {
+		if updateValidation != nil {
+			err := updateValidation(ctx, newRuntimeObj, oldRuntimeObj)
+			if err != nil {
+				klog.Infof("update failed validation: %v", err)
+				return nil, false, err
+			}
+		}
+
+		fieldErrors := r.updateStrategy.ValidateUpdate(ctx, newRuntimeObj, oldRuntimeObj)
+		if len(fieldErrors) > 0 {
+			return nil, false, apierrors.NewInvalid(api.SchemeGroupVersion.WithKind("PackageRevision").GroupKind(), name, fieldErrors)
 		}
 	}
 
-	fieldErrors := r.updateStrategy.ValidateUpdate(ctx, newRuntimeObj, oldObj)
-	if len(fieldErrors) > 0 {
-		return nil, false, apierrors.NewInvalid(api.SchemeGroupVersion.WithKind("PackageRevision").GroupKind(), oldObj.Name, fieldErrors)
+	if isCreate {
+		if createValidation != nil {
+			err := createValidation(ctx, newRuntimeObj)
+			if err != nil {
+				klog.Infof("update failed create validation: %v", err)
+				return nil, false, err
+			}
+		}
+
+		// TODO: ValidateCreate function ?
+		// fieldErrors := r.updateStrategy.ValidateCreate(ctx, newRuntimeObj, oldRuntimeObj)
+		// if len(fieldErrors) > 0 {
+		// 	return nil, false, apierrors.NewInvalid(api.SchemeGroupVersion.WithKind("PackageRevision").GroupKind(), name, fieldErrors)
+		// }
```

Comment:
or, like core k8s we should have CreateStrategy interface
[context](https://github.com/kubernetes/apiserver/blob/master/pkg/registry/rest/create.go#L42)
[k8s code](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L564-L575)

Reply:
That's #3090 so I'm going to merge this one!